### ISSUE TITLE
Modernize Database layer

### ISSUE THREAD

**Issue by tdesveaux** [buildbot/buildbot#7591]
As per #7385 , I'm adding type information here and there as I work on other things.

One thing I'm noticing is that the database models do not have [mapped python](https://docs.sqlalchemy.org/en/14/tutorial/metadata.html#declaring-mapped-classes) classes to interact, which would help typing. Is that by choice or a legacy implementation?

If I understand the code correctly, database data is passed around as `dict` which then need tests to validate that the data is as expected with `Validator` (defined [here](https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/util/validation.py#L294))?

Some things that could be done from the top of my head:
- require [sqlalchemy >=1.4](https://github.com/buildbot/buildbot/blob/master/master/setup.py#L660) (first released on [march 2021](https://docs.sqlalchemy.org/en/20/changelog/changelog_14.html#change-1.4.0)) which is the version made to be a [transition point to SQLAlchemy 2.0](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html) (this would allow to fix deprecation warnings ignore [here](https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/__init__.py#L163-L170))
- Implement DB models as mapped classes (or make dataclasses first, then make these dataclasses mapped afterward using [Hybrid Declaration](https://docs.sqlalchemy.org/en/14/orm/dataclasses.html#orm-declarative-dataclasses))

Also, it seems DB queries are run from a threadpool, an obvious reason is that queries are synchronous. If there is no other reason, SQLAlchemy 1.4.3 (March 2021) introduced asyncio support at a [beta level](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#) (although there is bugfixes as recently as a critical one in [1.4.51 (Jan 2024)](https://docs.sqlalchemy.org/en/20/changelog/changelog_14.html#change-13dac36b3e17cdbe4099c761e5151fb8), so there is reason to hold off on it.

Here is as far as I got looking into it for now. As I don't have a good understanding of Buildbot's internals nor history, I wanted to get maintainers opinion on this. I'm willing to submit PRs on this (hopefully, finding a way to make them as small as possible).

**Comment by p12tic** [comment]
I think what you wrote is mostly correct.

Dataclasses is a welcome addition.

One problem with asyncio is that Twisted doesn't work well with it and needs hacks, at least the last time I checked. Also, last time I checked there's no way to write asyncio tests that mock the clock. So for now I would like to not introduce asyncio as much as possible. 

So about your suggestions:

> require [sqlalchemy >=1.4](https://github.com/buildbot/buildbot/blob/master/master/setup.py#L660) (first released on [march 2021](https://docs.sqlalchemy.org/en/20/changelog/changelog_14.html#change-1.4.0)) which is the version made to be a [transition point to SQLAlchemy 2.0](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html) (this would allow to fix deprecation warnings ignore [here](https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/__init__.py#L163-L170))

Seems great, please go ahead.

> Implement DB models as mapped classes (or make dataclasses first, then make these dataclasses mapped afterward using [Hybrid Declaration](https://docs.sqlalchemy.org/en/14/orm/dataclasses.html#orm-declarative-dataclasses))

Also seems great. I think what you did in #7598 is the way to go. I just need to think a bit about migrations because this is part of public API. Is it possible to override `[]` operator to dataclass and provide an interface just like a dict?

**Comment by tdesveaux** [comment]
> Is it possible to override [] operator on a dataclass and provide an interface just like a dict? This way any existing third-party code would still work albeit with a lot of warnings.

Absolutely. In fact, it's what I did in the beginning before removing it in the last commit.
It's a simple:
```python
def __getitem__(self, key: str):
        if hasattr(self, key):
            return getattr(self, key)
        raise KeyError(key)
```

I'll remove the commit, and maybe add a deprecation warning?

As for the AsyncEngine, I think it's perfectly fine to postpone this. Current implementation using threads is perfectly fine, so using async would be of little gain.

**Comment by p12tic** [comment]
> I'll remove the commit, and maybe add a deprecation warning?

Yes, please do.
